### PR TITLE
Implement direct livestreaming

### PR DESCRIPTION
**Pull request by ktprograms**
## PR Checklist

Please check all that apply to this PR using "x":

- [x] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [x] I have checked that this PR does not introduce a breaking change

## PR Type

What kind of change does this PR introduce?

- [x] Feature

## Fixes

Issue Number: #94 

## What is the new behavior?

Direct livestreaming available

Note that currently you need to navigate to the go live fragment again after granting the audio permission in order for the camera permission to be requested.

### CODE REVIEW

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (kekkyojin) [inline app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:312]: This method will serve better if moved to Helper class. There are two other places where this same code is used and it is not depending on any Android framework variable or method. You could also move the other usages I mentioned to use the one from the Helper class.

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (kekkyojin) [inline app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:317]: My own personal preference is using curly brackets everywhere, even when there is only a single line of code in the branch.

No adding curly brackets could make bugs introduced if someone just adds a new line of code without realizing there are no curly lines of code. Compiling will fail then because of the else, but if there is no else, a potential bug could be introduced.

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (kekkyojin) [inline app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:324]: Add a new line at the end of every file

File: app/build.gradle
Comment (kekkyojin) [inline app/build.gradle:120]: Why this?

**Review by kekkyojin — COMMENTED** [review]
These are just comments. As always, not a formal review

File: app/build.gradle
Comment (ktprograms) [reply app/build.gradle:120]: [HaishinKit](https://github.com/shogo4405/HaishinKit.kt/) is a library for RTMP streaming for Android. I had to make a few changes in order to get it to build with Jitpack, rather than using the version from GitHub packages, which requires an authentication token. The other 3 dependencies are needed but don't seem to be pulled in by HaishinKit.

As I said, this is extremely WIP and I'm just making my current code available.

**Comment by ktprograms** [comment]
All three suggestions done in https://github.com/OdyseeTeam/odysee-android/pull/167/commits/40ec4b84af14f2488e1c6c6f5fec93a60547185f

(I'll squash the commits once it's ready for review)

**Comment by kekkyojin** [comment]
Thanks, @ktprograms

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (ktprograms) [inline app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:250]: Is hardcoding this okay, or should it be user specified?

File: app/src/main/res/layout/fragment_go_live.xml
Comment (kekkyojin) [inline app/src/main/res/layout/fragment_go_live.xml:84]: could you add `tools:text=""`with some sample error message? At design time, developers could change the `tools:visibility` to gone, as you have already added it

Comment (ktprograms) [reply app/src/main/res/layout/fragment_go_live.xml:85]: Done in https://github.com/OdyseeTeam/odysee-android/pull/167/commits/7bc99fa92a4340aeccef993c32d0817676efa2c1

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (ktprograms) [inline app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:339]: Any suggestions how to fix this?

I looked and it's not being caused by any calls to `setDisplayHomeAsUpEnabled` in the Odysee codebase.

**Comment by ktprograms** [comment]
Current status: It's ready for review but should not be merged because the commits aren't squashed, and due to a bug(?) in HaishinKit.kt. Once it's approved I'll squash the commits.

Blocker: The livestream does not have the correct aspect ratio. https://github.com/shogo4405/HaishinKit.kt/issues/48

@akinwale @kekkyojin Can you please review?

**Comment by kekkyojin** [comment]
The livestreaming feature doesn’t seem to need to offer additional UI changes, like showing other fragments along the livestream UI. Chat can be displayed in a layout. Doesn’t make sense implementing this on its own activity?

File: app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java
Comment (kekkyojin) [reply app/src/main/java/com/odysee/app/ui/golive/GoLiveFragment.java:339]: Absolutely not your fault. I have already seen this problem with other fragments, like the channel manager fragment.

**Comment by kekkyojin** [comment]
> Current status: It's ready for review but should not be merged because the commits aren't squashed, and due to a bug(?) in HaishinKit.kt. Once it's approved I'll squash the commits.
> 
There was a time when I suggested contributors to squash their commits. However, as GitHub is already offering this as an option when merging, squashing is no longer needed

**Comment by ktprograms** [comment]
>> Current status: It's ready for review but should not be merged because the commits aren't squashed, and due to a bug(?) in HaishinKit.kt. Once it's approved I'll squash the commits.
>
> There was a time when I suggested contributors to squash their commits. However, as GitHub is already offering this as an option when merging, squashing is no longer needed

I'd still prefer to do it manually, since the GitHub squash and merge puts all the commit messages into the merge commit's description.

**Comment by kekkyojin** [comment]
But it asks for confirmation and you could remove all the commits from the description, so only the commit message remains

**Comment by ktprograms** [comment]
> The livestreaming feature doesn’t seem to need to offer additional UI changes, like showing other fragments along the livestream UI. Chat can be displayed in a layout. Doesn’t make sense implementing this on its own activity?

Do you mean make livestreaming be in an Activity so it's easier to add features later on?

**Comment by ktprograms** [comment]
The bug in HaishinKit has been partially fixed. It seems to work fine if I don't change the cameraSource's resolution, but it still gets stretched with other resolutions. So the stream will be locked to 640x360px.

What should the bitrate be set to, and should it be user configurable? The default in HaishinKit is 500kbit/s.

**Comment by ktprograms** [comment]
Ok, I see. That makes sense.

Just FYI there isn't any chat currently.

**Comment by ktprograms** [comment]
@kekkyojin Should permissions be handled in the GoLiveActivity or use the callbacks in MainActivity?

**Comment by ktprograms** [comment]
Ok, moved livestreaming into an activity.

(Did the permission handling directly in the activity)

File: app/src/main/java/com/odysee/app/GoLiveActivity.java
Comment (kekkyojin) [inline app/src/main/java/com/odysee/app/GoLiveActivity.java:596]: This preference is now a little bit more complex as it could be "Use system setting". Check the implementation on one of the latest commits

File: app/src/main/java/com/odysee/app/GoLiveActivity.java
Comment (kekkyojin) [inline app/src/main/java/com/odysee/app/GoLiveActivity.java:564]: Use ClaimList supplier for API Level 24+

Comment (ktprograms) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java:564]: I noticed that, but there's quite a lot of duplicated code by doing that, right?

What's the difference between the two?

Comment (kekkyojin) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java:564]: ClaimListTask is an AsyncTask, which is deprecated. New code should not be using AsyncTask. ClaimListTask remains on repo because there is still code using it.

You could just use the supplier forAPI Level 24+, leaving the AsyncTask to older API Levels.

File: app/src/main/java/com/odysee/app/GoLiveActivity.java
Comment (kekkyojin) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java:596]: See getDarkModeSetting() method on MainActivity.

For Android 10+, dark mode could be Light/Night/System, being "Light" the default one; this setting is a string. On previous versions, it is a boolean for light/dark and default is "false" -meaning Light-.

Comment (ktprograms) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java:596]: Yes, I found that code, but what I'm saying is that the UI seems to handle light/dark mode fine even without the code at https://github.com/OdyseeTeam/odysee-android/pull/167/files#diff-d31a44933ced9a1e3d8d3fd3123fbc7607c595729a97d9daf98e0e63cb4f0aadR114-R117

File: app/src/main/java/com/odysee/app/GoLiveActivity.java
Comment (ktprograms) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java]: Ok, done in https://github.com/OdyseeTeam/odysee-android/pull/167/commits/b219673e1bc5a589484818151d3202b58793b23a

**Comment by ktprograms** [comment]
@akinwale Ok, done.

Few discussion points:
- Currently no alert is shown when pressing back while confirming, which would make the user lose access to that livestream session.
- I've just made it not reset the livestream flow when resuming while waiting for confirmation, but it won't save any state, so when the GoLiveActivity instance gets destroyed it won't be able to resume.

**Comment by ktprograms** [comment]
Rebased against master.
=

**Review by akinwale — APPROVED** [review]
Looks good. Thanks.

File: app/src/main/java/com/odysee/app/GoLiveActivity.java
Comment (ktprograms) [reply app/src/main/java/com/odysee/app/GoLiveActivity.java:679]: @kekkyojin Is it necessary to add the light/dark/system handling code to GoLiveActivity?

The UI (including the status bar) seems to adapt properly even when I switch to dark mode while the app is running.

**Pull request merged**
